### ISSUE TITLE
Internal Frames do not have decorations corresponding to FlatLaf

### ISSUE THREAD

**Issue by Chrriis** [JFormDesigner/FlatLaf#39]
(bug report split from issue #11)

Internal frame decorations are not correct for FlatLaf.
Here is a simple test case:
```Java
    public static void main(String[] args) throws Exception {
        UIManager.setLookAndFeel(new FlatLightLaf());
        JFrame frame = new JFrame();
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        JDesktopPane desktopPane = new JDesktopPane();
        JInternalFrame internalFrame = new JInternalFrame("Test", true, true, true, true);
        internalFrame.setLocation(50, 30);
        internalFrame.setSize(300, 200);
        desktopPane.add(internalFrame);
        frame.getContentPane().add(desktopPane);
        frame.setSize(400, 300);
        frame.setVisible(true);
        internalFrame.setVisible(true);
    }
```

And here is the result:
![InternalFrame](https://user-images.githubusercontent.com/1174202/72105333-aa2f6880-332d-11ea-9072-684cf40f8ba4.png)

**Comment by Chrriis** [comment]
@DevCharly You said that you started some work on internal frames. Could you share that work in progress? Internal frame support is a show stopper for us so I would take anything you may have 😉

**Comment by Chrriis** [comment]
Modern applications that use flat widgets may want their frame decorations also to be flat. Here, I am generalizing the idea to JFrame, not limiting to JInternalFrame.
I assume that the implementation for internal frames could be leveraged for frames as well.

Here is a test case of the current situation:
```Java
	public static void main(String[] args) throws Exception {
		UIManager.setLookAndFeel(new FlatLightLaf());
		JFrame frame = new JFrame("JFrame");
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frame.setUndecorated(true);
		frame.getRootPane().setWindowDecorationStyle(JRootPane.FRAME);
		JDesktopPane desktopPane = new JDesktopPane();
		JInternalFrame internalFrame1 = new JInternalFrame("JInternalFrame1", true, true, true, true);
		internalFrame1.setLocation(50, 30);
		internalFrame1.setSize(300, 200);
		desktopPane.add(internalFrame1);
		JInternalFrame internalFrame2 = new JInternalFrame("JInternalFrame2", true, true, true, true);
		internalFrame2.setLocation(60, 40);
		internalFrame2.setSize(300, 200);
		desktopPane.add(internalFrame2);
		frame.getContentPane().add(desktopPane);
		frame.setSize(400, 300);
		frame.setVisible(true);
		internalFrame1.setVisible(true);
		internalFrame2.setVisible(true);
		internalFrame2.setIcon(true);
	}
```

... and here is the result:
![FrameDecorations](https://user-images.githubusercontent.com/1174202/72375936-4fbb5100-370d-11ea-872a-d1eab44b5cdf.png)

**Comment by DevCharly** [comment]
Window decoration style looks similar to internal frames, but it seems that the implementation has nothing to do with internal frames. It is implemented in `javax.swing.plaf.metal.MetalRootPaneUI`and `javax.swing.plaf.metal.MetalTitlePane`. But I keep that in mind and hope that it is possible to share some code...

Could you please open another issue?

**Comment by DevCharly** [comment]
> You said that you started some work on internal frames

I've only implemented a [test application](https://github.com/JFormDesigner/FlatLaf/blob/master/flatlaf-testing/src/main/java/com/formdev/flatlaf/testing/FlatInternalFrameTest.java) and four icons for the internal frame header (e.g. [close icon](https://github.com/JFormDesigner/FlatLaf/blob/master/flatlaf-core/src/main/java/com/formdev/flatlaf/icons/FlatInternalFrameCloseIcon.java)).

Plan to work on this next week...

**Comment by DevCharly** [comment]
@rogerbj @Chrriis 
A basic implementation of `JInternalFrame` is now in master branch. `JDesktopPane` is not yet implemented.

![image](https://user-images.githubusercontent.com/5604048/73185104-c7ca4380-411d-11ea-8a35-8e312b50ce4f.png)

![image](https://user-images.githubusercontent.com/5604048/73185112-cbf66100-411d-11ea-8d78-5acccd22e37e.png)

To make resizing easy, there is an invisible 5px border around the internal frames. This means that the visual bounds are slightly smaller than the frame bounds. Hope this does not make problems...

Please try it out and l report any issues. Thx.

**Comment by Chrriis** [comment]
@DevCharly from what I have tested, it seems to work well! Thanks a lot!

I find that the buttons are a bit small. When moving the mouse, it is harder to aim, and clicking close instead of maximize can be frustrating. Maybe they should have a bigger margin (or simply be wider).
Here is a comparison with Windows L&F:
![InternalFrameWinLafvsFlatLaf](https://user-images.githubusercontent.com/1174202/73219587-d126d080-415c-11ea-9b41-ede93452d5fa.png)

About the empty border, resizing would be a pain without it.
Minor: I noticed that while hovering the empty border you get the resize cursor (good), but if you click (and release) without moving the mouse then the cursor changes to arrow, although it is still over the empty border.

**Comment by DevCharly** [comment]
@Chrriis thanks for your feedback.

I've made the buttons larger. That's better!

![image](https://user-images.githubusercontent.com/5604048/73693103-e27c5980-46d5-11ea-8bdb-beb202163625.png)

**Comment by Chrriis** [comment]
@DevCharly It is perfect! Good job!

**Comment by DevCharly** [comment]
@Chrriis @rogerbj don't know whether you use `JDesktopPane` for iconified internal frames. It is now implemented in master branch.

Iconified internal frames are show as a simple square icons that contains a preview of the internal frame. Clicking the icon restores the internal frame. The frame title and a close button are shown when the mouse pointer is over an icon.

![image](https://user-images.githubusercontent.com/5604048/74537860-2fb0c480-4f3b-11ea-82a9-b510aa967af5.png)

**Comment by Chrriis** [comment]
@DevCharly I don't use iconified internal frames, but I gave it a try to give you some feedback.

I found something surprising: if I iconify Frame1 and Frame2, then Frame1 is on the left followed by Frame2. Then I restore both and iconify Frame2: it does not get iconified on the left, but retains its previous iconified position. I expected to take the first empty slot starting from the left. Otherwise, what if someone has 10 frames and iconifies the 10th one only? Would it appear in the middle of the screen? 🙂

**Comment by DevCharly** [comment]
Thanks for testing 👍 

That behavior is indeed a little bit strange. This is implemented somewhere in Swing. Windows, Metal and Nimbus LaF behave exactly the same. Have not checked whether it is possible to change this...

**Comment by kirill-grouchnikov** [comment]
The caching is probably in the `DefaultDesktopManager.iconifyFrame`

**Comment by DevCharly** [comment]
Thanks @kirill-grouchnikov . I'll check this.

**Comment by DevCharly** [comment]
implemented in 0.27
